File: fabric-audit-agent/core/automation/trend.js
/**
 * Annotate each finding with how many of the recent runs (window) contained its key.
 * Adds `recurringRuns` (integer, current run counts as 1). Pure.
 * @param {object[]} findings
 * @param {Array<{findings:Array<{key:string}>}>} history chronological
 * @param {number} window default 7
 * @returns {object[]}
 */
export function annotateRecurring(findings, history, window = 7) {
  const recent = history.slice(-window);
  return findings.map(f => {
    const priorHits = f.key ? recent.filter(run => run.findings.some(rf => rf.key === f.key)).length : 0;
    return { ...f, recurringRuns: priorHits + 1 };
  });
}

File: fabric-audit-agent-py/fabric_audit_agent/automation/trend.py
"""Recurrence trend. Port of ``core/automation/trend.js``.

Annotate each finding with how many recent runs (window) contained its key, as
``recurringRuns`` (current run counts as 1). Pure.
"""


def annotate_recurring(findings, history, window=7):
    recent = history[-window:]
    out = []
    for f in findings:
        prior_hits = sum(1 for run in recent if any(rf.get("key") == f["key"] for rf in run["findings"])) if f.get("key") else 0
        out.append({**f, "recurringRuns": prior_hits + 1})
    return out
